### ISSUE TITLE
More Visma Problems

### ISSUE THREAD

**Issue by macasas** [tamas-p/hisma#24]
Hi tamas, apologies for writing again but this is driving me crazy.

Here is my project https://github.com/macasas/ssv2 

It was developing fine, visma displaying fine, and then suddenly yesterday, just as all the machines and routes were pointing in the right direction, visma just disconnected while app was in progress, and then wouldn't reconnect.  Now it is connecting but only going as far as the SMWatcher.

I've taken my app all the way back to just the app and no regions, routing to 3 screens (currently commented out) and visma does not show my machines.

I would very much appreciate you taking a look because I feel something in my code is causing this to not work. Perhaps something obvious that I just cannot see. The app is now actually routing as designed, I just cannot see it happening in visma :-(

Thanks

Using the following has worked up until yesterday.
Browser at http://127.0.0.1:4040/
visma --port=4040 --plantuml_jar=plantuml.1.2022.1.jar --bind 0.0.0.0

**Comment by tamas-p** [comment]
Hello, I will have a look in the evening (CET).

**Comment by tamas-p** [comment]
One quick note, that in your main() at https://github.com/macasas/ssv2 you are are trying to connect to port 4030:

```dart
  StateMachine.monitorCreators = [
    (m) => VisualMonitor(
      m,
      // host: 'localhost',
      port: 4030,
    ),
    (m) => ConsoleMonitor(m),
  ];
```
and your visma service is running on 4040:
```console
visma --port=4040 --plantuml_jar=plantuml.1.2022.1.jar --bind 0.0.0.0
```
This results exactly what you see, that VisualMonitor is not able to connect visma (this would be visible in the logs with some log configuration) and visma is just sitting there waiting empty and presenting you only the empty SM Watcher.

**Comment by macasas** [comment]
I will check the 4030 but this was changed after the problem appeared trying to get it working. I also have some time later today yo take another look, thanks

**Comment by macasas** [comment]
Here is an explanation, I feel I am doing something stupid, but I just cannot see what that is. :-(

[youtube](https://youtu.be/oKhrIZAQxkE)

**Comment by tamas-p** [comment]
I've managed to reproduce the issue with the code you provided. I hope to come back with the root cause soon.

**Comment by tamas-p** [comment]
Well, the actual root cause is simple: In https://github.com/macasas/ssv2/blob/main/lib/src/app/regions/account_status/layers/base/account_status_machine.dart there is transition referenced - but it is not defined:

```dart
          AccSTmE.evCancelledRegistration: [AccSTmT.toCancelledRegistration],
```
that is in the accountStatus state definition.
When I commented out that `event to transition` entry it started working immediately. 
```dart
    states: {
      AccSTmS.accountStatus: State(
        etm: {
          AccSTmE.noaccount: [AccSTmT.toNoAccount],
          AccSTmE.account: [AccSTmT.toAccount],
         // AccSTmE.evCancelledRegistration: [AccSTmT.toCancelledRegistration],
        },
```
Instead of commenting it out, If you define that transition to your need, visma shall work.

Surely visma - more specifically visual_monitor - shall be more robust by either providing the precise error message what causing the problem or being more permissive and simply not draw that transition that is however referenced in an `etm` but not defined. Actually there is already an assert for this in visual_monitor, but I made a mistake in websocket connection loop that retries connection if an exception was caught: It is not logged out. I will add that line to log this out and release the package.

I tend to think that an error/exception is better versus being permissive in visual_monitor. Furthermore, that check shall be in the base hisma package (there's already an assert, but it only steps in if you try firing that event). That way one could catch early such problems even without using visual_monitor. This might be a health check of the machine in the state machine constructor.

**Comment by tamas-p** [comment]
I committed the fix to master: 403ddcfdd99ef52b45dd199b39edd47ec51d8c7a
Now the assert is not caught hence it aborts the program:
```console
Unhandled exception:
'package:hisma_visual_monitor/src/plantuml/plantuml_converter.dart': Failed assertion: line 619 pos 11: 'transition != null': In machine named 'lightMachine' at state 'S.off' for event 'E.stop' the transition referenced by 'T.quickTest' is not defined.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:51:61)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
#2      MachineConverter._writeTransitions.<anonymous closure> (package:hisma_visual_monitor/src/plantuml/plantuml_converter.dart:619:11)
#3      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:633:13)
#4      MachineConverter._writeTransitions (package:hisma_visual_monitor/src/plantuml/plantuml_converter.dart:615:15)
#5      MachineConverter._iterateStates.<anonymous closure> (package:hisma_visual_monitor/src/plantuml/plantuml_converter.dart:237:27)
#6      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:633:13)
#7      MachineConverter._iterateStates (package:hisma_visual_monitor/src/plantuml/plantuml_converter.dart:236:20)
#8      MachineConverter._convert (package:hisma_visual_monitor/src/plantuml/plantuml_converter.dart:175:39)
#9      new MachineConverter (package:hisma_visual_monitor/src/plantuml/plantuml_converter.dart:99:5)
#10     PlantUMLConverter.diagram (package:hisma_visual_monitor/src/plantuml/plantuml_converter.dart:43:7)
#11     VisualMonitor._sendRegistration (package:hisma_visual_monitor/src/visual_monitor/client/visual_monitor.dart:128:7)
#12     VisualMonitor._tryConnectCord (package:hisma_visual_monitor/src/visual_monitor/client/visual_monitor.dart:109:5)

```
Also, if run without assertions enabled the PlantUMLConverer is permissive and will simply not draw that transition that is not defined. Release of the package will follow soon.

**Comment by tamas-p** [comment]
Released in https://pub.dev/packages/hisma/versions/0.3.2+1
